arch/riscv/core/pmp/core_pmp.c

```diff
@@ -0,0 +1,223 @@
+	CSR_PMPADDR5,
+	CSR_PMPADDR6,
+	CSR_PMPADDR7,
+	CSR_PMPADDR8,
+	CSR_PMPADDR9,
+	CSR_PMPADDR10,
+	CSR_PMPADDR11,
+	CSR_PMPADDR12,
+	CSR_PMPADDR13,
+	CSR_PMPADDR14,
+	CSR_PMPADDR15
+};
+
+ulong_t csr_read_enum(int pmp_csr_enum)
+{
+	ulong_t res = -1;
+
+	switch (pmp_csr_enum) {
+	case CSR_PMPCFG0:
+		res = csr_read(0x3A0); break;
+	case CSR_PMPCFG1:
+		res = csr_read(0x3A1); break;
+	case CSR_PMPCFG2:
+		res = csr_read(0x3A2); break;
+	case CSR_PMPCFG3:
+		res = csr_read(0x3A3); break;
+	case CSR_PMPADDR0:
+		res = csr_read(0x3B0); break;
+	case CSR_PMPADDR1:
+		res = csr_read(0x3B1); break;
+	case CSR_PMPADDR2:
+		res = csr_read(0x3B2); break;
+	case CSR_PMPADDR3:
+		res = csr_read(0x3B3); break;
+	case CSR_PMPADDR4:
+		res = csr_read(0x3B4); break;
+	case CSR_PMPADDR5:
+		res = csr_read(0x3B5); break;
+	case CSR_PMPADDR6:
+		res = csr_read(0x3B6); break;
+	case CSR_PMPADDR7:
+		res = csr_read(0x3B7); break;
+	case CSR_PMPADDR8:
+		res = csr_read(0x3B8); break;
+	case CSR_PMPADDR9:
+		res = csr_read(0x3B9); break;
+	case CSR_PMPADDR10:
+		res = csr_read(0x3BA); break;
+	case CSR_PMPADDR11:
+		res = csr_read(0x3BB); break;
+	case CSR_PMPADDR12:
+		res = csr_read(0x3BC); break;
+	case CSR_PMPADDR13:
+		res = csr_read(0x3BD); break;
+	case CSR_PMPADDR14:
+		res = csr_read(0x3BE); break;
+	case CSR_PMPADDR15:
+		res = csr_read(0x3BF); break;
+	default:
+		break;
+	}
+	return res;
+}
+
+void csr_write_enum(int pmp_csr_enum, ulong_t value)
+{
+	switch (pmp_csr_enum) {
+	case CSR_PMPCFG0:
+		csr_write(0x3A0, value); break;
+	case CSR_PMPCFG1:
+		csr_write(0x3A1, value); break;
+	case CSR_PMPCFG2:
+		csr_write(0x3A2, value); break;
+	case CSR_PMPCFG3:
+		csr_write(0x3A3, value); break;
+	case CSR_PMPADDR0:
+		csr_write(0x3B0, value); break;
+	case CSR_PMPADDR1:
+		csr_write(0x3B1, value); break;
+	case CSR_PMPADDR2:
+		csr_write(0x3B2, value); break;
+	case CSR_PMPADDR3:
+		csr_write(0x3B3, value); break;
+	case CSR_PMPADDR4:
+		csr_write(0x3B4, value); break;
+	case CSR_PMPADDR5:
+		csr_write(0x3B5, value); break;
+	case CSR_PMPADDR6:
+		csr_write(0x3B6, value); break;
+	case CSR_PMPADDR7:
+		csr_write(0x3B7, value); break;
+	case CSR_PMPADDR8:
+		csr_write(0x3B8, value); break;
+	case CSR_PMPADDR9:
+		csr_write(0x3B9, value); break;
+	case CSR_PMPADDR10:
+		csr_write(0x3BA, value); break;
+	case CSR_PMPADDR11:
+		csr_write(0x3BB, value); break;
+	case CSR_PMPADDR12:
+		csr_write(0x3BC, value); break;
+	case CSR_PMPADDR13:
+		csr_write(0x3BD, value); break;
+	case CSR_PMPADDR14:
+		csr_write(0x3BE, value); break;
+	case CSR_PMPADDR15:
+		csr_write(0x3BF, value); break;
+	default:
+		break;
+	}
+}
+
+int z_riscv_pmp_set(unsigned int index, ulong_t cfg_val, ulong_t addr_val)
+{
+	ulong_t reg_val;
+	ulong_t shift, mask;
+	int pmpcfg_csr;
+	int pmpaddr_csr;
+
+	if ((index >= PMP_SLOT_NUMBER) | (index < 0))
+		return -1;
+
+	/* Calculate PMP config/addr register, shift and mask */
+#ifdef CONFIG_64BIT
+	pmpcfg_csr = CSR_PMPCFG0 + (index >> 4);
```

Comment:
I think the pmpcfg index is `((index >> 3) << 1)` in RV64 because pmp[0 ~ 7]cfg maps to cfg0 and pmp[8 ~ 15]cfg maps to cfg2.

Reply:
I'm sorry but I don't understand your fix and explanation.
Why CSR_PMPCFG0 + (index >> 4) is wrong ? As you said: "pmp[0 ~ 7]cfg maps to cfg0 and pmp[8 ~ 15]cfg maps to cfg2"
It seems do the job, isn't it ?
I'm probably wrong, but I don't understand your fix. Can you re-explain me please ?
Thanks !

Reply:
Thanks for your response.
I think it's wrong because using `(index >> 4)` will map all pmp entries (0 ~ 15) to pmpcfg0 CSR. (If I understand code correctly)
Using `(index / 8) * 2`, equals to `((index >> 3) << 1)` will map pmp entry 0 ~ 7 to cfg0 and pmp entry 8 ~ 15 to cfg2.

Reply:
Understood, I shift too much. Then "CSR_PMPCFG0 + (index >> 3)" should be right I think.
Add "<< 1" could be spawn 1 or 0 on LSB depend on CPU. (And I think it's 0)